tests/suite/upgrade_test.go

```diff
@@ -0,0 +1,246 @@
+		type metricsResults struct {
+			metrics  *vegeta.Metrics
+			testName string
+			scheme   string
+		}
+		metricsCh := make(chan *metricsResults)
+
+		type testCfg struct {
+			desc   string
+			port   string
+			target framework.Target
+		}
+
+		tests := []testCfg{
+			{
+				desc: "Send http /coffee traffic",
+				port: "80",
+				target: framework.Target{
+					Method: "GET",
+					URL:    "http://cafe.example.com/coffee",
+				},
+			},
+			{
+				desc: "Send https /tea traffic",
+				port: "443",
+				target: framework.Target{
+					Method: "GET",
+					URL:    "https://cafe.example.com/tea",
+				},
+			},
+		}
+
+		for _, test := range tests {
+			go func(cfg testCfg) {
+				defer GinkgoRecover()
+
+				results, metrics := framework.RunLoadTest(
+					[]framework.Target{cfg.target},
+					100,
+					60*time.Second,
+					cfg.desc,
+					fmt.Sprintf("%s:%s", address, cfg.port),
+					"cafe.example.com",
+				)
+
+				scheme := strings.Split(cfg.target.URL, "://")[0]
+				metricsRes := metricsResults{
+					metrics:  &metrics,
+					testName: fmt.Sprintf("\n## Test: %s\n\n```text\n", cfg.desc),
+					scheme:   scheme,
+				}
+
+				buf := new(bytes.Buffer)
+				encoder := vegeta.NewCSVEncoder(buf)
+				for _, res := range results {
+					res := res
+					Expect(encoder.Encode(&res)).To(Succeed())
+				}
+
+				csvName := fmt.Sprintf("%s.csv", scheme)
+				filename := filepath.Join(resultsDir, csvName)
+				csvFile, err := framework.CreateResultsFile(filename)
+				Expect(err).ToNot(HaveOccurred())
+
+				_, err = fmt.Fprint(csvFile, buf.String())
+				Expect(err).ToNot(HaveOccurred())
+				csvFile.Close()
+
+				output, err := framework.GeneratePNG(resultsDir, csvName, fmt.Sprintf("%s.png", scheme))
+				Expect(err).ToNot(HaveOccurred(), string(output))
+
+				metricsCh <- &metricsRes
+			}(test)
+		}
+
+		// allow traffic flow to start
+		time.Sleep(2 * time.Second)
+
+		output, err := framework.UpgradeNGF(cfg, "--values", valuesFile)
+		Expect(err).ToNot(HaveOccurred(), string(output))
+
+		Expect(resourceManager.ApplyFromFiles([]string{"ngf-upgrade/gateway-updated.yaml"}, ns.Name)).To(Succeed())
+
+		podNames, err := framework.GetNGFPodNames(k8sClient, ngfNamespace, releaseName, timeoutConfig.GetTimeout)
+		Expect(err).ToNot(HaveOccurred())
+
+		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+		defer cancel()
+
+		// ensure that the leader election lease has been updated to the new pods
+		var lease coordination.Lease
+		key := types.NamespacedName{Name: "ngf-test-nginx-gateway-fabric-leader-election", Namespace: ngfNamespace}
+		Expect(k8sClient.Get(ctx, key, &lease)).To(Succeed())
+
+		Expect(lease.Spec.HolderIdentity).ToNot(BeNil())
+		Expect(podNames).To(ContainElement(*lease.Spec.HolderIdentity))
+
+		// ensure that the Gateway has been properly updated with a new listener
+		var gw v1.Gateway
+		key = types.NamespacedName{Name: "gateway", Namespace: ns.Name}
+		Expect(wait.PollUntilContextCancel(
+			ctx,
```

Comment:
I wonder if we should use a separate context here so each lookup has its own timeout. What do you think?